File: python/src/storylet_framework/context.py
# This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
# Copyright (c) 2025 Ian Thomas

import sys
import os
from typing import Any, Dict, List, Optional

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../lib/expression-parser")))

from expression_parser.parser import Parser

expression_parser = Parser()

def eval_expression(val: Any, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Any:
    """
    Use the ExpressionParser to evaluate an expression
    Unless it's a simple bool or number, in which case just use that.
    dump_eval will fill an array with evaluation debug steps
    """
    if isinstance(val, (bool, int, float)):
        return val
    expression = expression_parser.parse(val)
    return expression.evaluate(context, dump_eval)


def init_context(context: Dict[str, Any], properties: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> None:
    """
    Evaluate each expression in 'properties' and add it to the context.
    An error is raised if a property already exists in the context.
    """
    for prop_name, expr in properties.items():
        if dump_eval is not None:
            dump_eval.append(f"InitContext: Evaluating {prop_name} = {expr}")
        result = eval_expression(expr, context, dump_eval)
        if prop_name in context:
            raise ValueError(f"Trying to initialise property '{prop_name}' in context when it already exists.")
        context[prop_name] = result


def update_context(context: Dict[str, Any], updates: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> None:
    """
    Evaluate any entries in updates and apply them to the context, but they must already exist
    dump_eval will fill an array with evaluation debug steps
    """
    for prop_name, expr in updates.items():
        if dump_eval is not None:
            dump_eval.append(f"UpdateContext: Evaluating {prop_name} = {expr}")
        result = eval_expression(expr, context, dump_eval)
        #print(f"Setting {prop_name} to {result}");
        if prop_name not in context:
            raise KeyError(f"Context var: '{prop_name}' undefined.")
        context[prop_name] = result


def dump_context(context: Dict[str, Any]) -> str:
    """
    For debugging - dump out the vars in the context.
    """
    lines = []
    for prop_name, expr in context.items():
        if callable(expr):
            lines.append(f"{prop_name} = <function>")
        elif isinstance(expr, (bool, int, float, str)):
            result = eval_expression(expr, context)
            lines.append(f"{prop_name} = {result}")
            
    return "\n".join(lines)

File: csharp/StoryletFramework/Context.cs
/*
 * This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
 * Copyright (c) 2025 Ian Thomas
 */

using ExpressionParser;

namespace StoryletFramework;

using Context = Dictionary<string, object>;
using KeyMap = Dictionary<string, object>;

public static class ContextUtils
{
    private static readonly Parser expressionParser = new Parser();

    // Use the ExpressionParser to evaluate an expression
    // Unless it's a simple bool or number, in which case just use that.
    // dumpEval will fill a list with evaluation debug steps
    public static object EvalExpression(object val, Context context, List<string>? dumpEval = null)
    {
        if (val is bool || val is double || val is int)
            return val;

        if (val is string str) {
                var expression = expressionParser.Parse(str);
            return expression?.Evaluate(context, dumpEval) ?? throw new InvalidOperationException("Expression result should never be null.");
        }
        
        throw new ArgumentNullException(nameof(val), "Expression text cannot be null or empty.");
    }

    // Evaluate and copy any new entries in 'properties' into the context
    // dumpEval will fill a list with evaluation debug steps
    public static void InitContext(Context context, KeyMap properties, List<string>? dumpEval = null)
    {
        foreach (var kvp in properties)
        {
            var propName = kvp.Key;
            var expression = kvp.Value;

            if (dumpEval != null)
                dumpEval.Add($"InitContext: Evaluating {propName} = {expression}");

            var result = EvalExpression(expression, context, dumpEval);

            if (context.ContainsKey(propName))
                throw new InvalidOperationException($"Trying to initialize property '{propName}' in context when it already exists.");

            context[propName] = result;
        }
    }

    // Evaluate any entries in updates and apply them to the context, but they must already exist
    // dumpEval will fill a list with evaluation debug steps
    public static void UpdateContext(Context context, KeyMap updates, List<string>? dumpEval = null)
    {
        foreach (var kvp in updates)
        {
            var propName = kvp.Key;
            var expression = kvp.Value;

            if (dumpEval != null)
                dumpEval.Add($"UpdateContext: Evaluating {propName} = {expression}");

            var result = EvalExpression(expression, context, dumpEval);

            if (!context.ContainsKey(propName))
                throw new KeyNotFoundException($"Context variable '{propName}' is undefined.");

            context[propName] = result;
        }
    }

    // For debugging: Dump the context as a string
    public static string DumpContext(Context context)
    {
        var output = new List<string>();

        foreach (var kvp in context)
        {
            var propName = kvp.Key;
            var expression = kvp.Value;

            if (expression is Delegate)
            {
                output.Add($"{propName} = <function>");
            }
            else if (expression is bool || expression is double || expression is int || expression is string)
            {
                var result = EvalExpression(expression, context);
                output.Add($"{propName} = {result}");
            }
        }

        return string.Join("\n", output);
    }
}